android/src/main/java/com/guichaguri/trackplayer/service/player/IcyStreamUpdate.java

```diff
@@ -0,0 +1,52 @@
+package com.guichaguri.trackplayer.service.player;
+
+import android.util.Log;
+import android.os.Bundle;
+
+import com.guichaguri.trackplayer.module.MusicEvents;
+import com.guichaguri.trackplayer.service.MusicManager;
+import com.guichaguri.trackplayer.service.MusicService;
+import com.guichaguri.trackplayer.service.Utils;
+
+import saschpe.exoplayer2.ext.icy.*;
+
+
+public class IcyStreamUpdate {
+
+    private static MusicService service = null;
+
+    public static void setService(MusicService service) {
+        IcyStreamUpdate.service = service;
+    }
+    
+    public static IcyHttpDataSource.IcyHeadersListener HeaderUpdate = new IcyHttpDataSource.IcyHeadersListener() {
+        @Override
+        public void onIcyHeaders(IcyHttpDataSource.IcyHeaders icyHeaders) {
+            Log.d(Utils.LOG, "HeaderUpdate: " + icyHeaders.toString());  
+            if(IcyStreamUpdate.service != null) {
+                
+                Bundle bundle = new Bundle();
+                bundle.putInt("bitRate", icyHeaders.getBitRate());
+                bundle.putString("genre", icyHeaders.getGenre());
+                bundle.putString("name", icyHeaders.getName());
+                bundle.putString("url", icyHeaders.getUrl());
+                bundle.putBoolean("public", icyHeaders.isPublic());
+                IcyStreamUpdate.service.emit(MusicEvents.ICY_HEADER_UPDATE, bundle);   
+            }  
+        }
+    };
+    
+    public static IcyHttpDataSource.IcyMetadataListener MetaDataUpdate = new IcyHttpDataSource.IcyMetadataListener() {
+        @Override
+        public void onIcyMetaData(IcyHttpDataSource.IcyMetadata icyMetadata) {
+            Log.d(Utils.LOG, "MetaData Update! " + icyMetadata.toString());  
+            if(IcyStreamUpdate.service != null) {
+                
+                Bundle bundle = new Bundle();
+                bundle.putString("streamTitle", icyMetadata.getStreamTitle());
+                bundle.putString("streamUrl", icyMetadata.getStreamUrl());
+                IcyStreamUpdate.service.emit(MusicEvents.ICY_MATADATA_UPDATE, bundle);      
+            }    
+        }
+    };
+}
```

Comment:
This class seems totally unnecessary, it seems to much code to simply handle two events.
I also don't like the fact that it is static, as you can easily leak memory.

Reply:
I would really like to not use static variables, due to the problem you have mentioned yourself, but since I did not find a way to get the Service through a Track, I ended up doing this way, I can review the code and find other wiggles, in my GIT in a Brach I did not Push in which I do not use static variable, but I do not yet attribute the value of `service`.

Reply:
Look into `MusicManager`, it stores instances for a lot of classes, and this one should also belong there.